### ISSUE TITLE
Formalize public API for registering 3rd party cops

### ISSUE THREAD

**Issue by backus** [rubocop-hq/rubocop#3414]
I've been updating, refactoring, and adding to [`rubocop-rspec`](https://github.com/nevir/rubocop-rspec) recently. It works well but there are some major and minor pain points associated with maintaining a third party RuboCop gem. The major blemish is making configuration work:

``` ruby
module RuboCop
  module RSpec
    CONFIG_DEFAULT = PROJECT_ROOT.join('config', 'default.yml').freeze

    # Because RuboCop doesn't yet support plugins, we have to monkey patch in a
    # bit of our configuration.
    module Inject
      def self.defaults!
        path = CONFIG_DEFAULT.to_s
        hash = ConfigLoader.send(:load_yaml_configuration, path)
        config = Config.new(hash, path)
        puts "configuration from #{path}" if ConfigLoader.debug?
        config = ConfigLoader.merge_with_default(config, path)
        ConfigLoader.instance_variable_set(:@default_configuration, config)
      end
    end
  end
end
```

This ugliness is necessary to load configuration options into rubocop at require time without issuing "unrecognized cop" warnings.

---

In the long term I'd like to open separate issues detailing other third party awkwardness that should be resolved. In the short term (for this issue) I'd like to propose and discuss a smaller scope change to RuboCop that will make it easier to write third party gems like rubocop-rspec and also extract rubocop-rails. Once we agree on these changes I'd be happy to open PRs implementing the changes.

**Comment by backus** [comment]
My preliminary proposal is pretty simple. There should be a `--use` command line flag:

```
$ rubocop --help
Usage: rubocop [options] [file1, file2, ...]
    ...

    -u, --use GEM                    Load RuboCop extension gem.
```

Usage might look like this

```sh
rubocop --use rubocop-rspec --use rubocop-rails
```

This flag would be similar to `--require` but would also look for `config/default.yml` within the root of the required gem. This `config/default.yml` would be loaded just like RuboCop's main configuration is loaded and the results would be merged with RuboCop's default configuration.

**Comment by backus** [comment]
@bbatsov or @jonas054 could one of you either sign off on this proposal or share a counter-proposal if you don't like this? I am happy to do the work to make this happen but need to know that the work is wanted and in scope.

**Comment by bbatsov** [comment]
I'm fine with the proposed change.

**Comment by bbatsov** [comment]
> In the long term I'd like to open separate issues detailing other third party awkwardness that should be resolved. In the short term (for this issue) I'd like to propose and discuss a smaller scope change to RuboCop that will make it easier to write third party gems like rubocop-rspec and also extract rubocop-rails. Once we agree on these changes I'd be happy to open PRs implementing the changes.

I'd be happy to see someone driving this. It's quite important and the work to properly support 3rd-party add-ons is long overdue.

**Comment by tjwp** [comment]
I would have preferred something more automatic. Something along the lines of how cops are added to the `CopStore` when they inherit from `Rubocop::Cop`.

Having each third-party gem subclass something (introduce a `Rubocop::DefaultConfiguration`?), or register itself in some other way, would be easier to work with than a long command line if several extensions are used.

**Comment by Drenmi** [comment]
> This config/default.yml would be loaded just like RuboCop's main configuration is loaded and the results would be merged with RuboCop's default configuration.

I'm assuming the plugin defaults should take precedence, in case they need to override the config for existing cops?

**Comment by bbatsov** [comment]
I'll close this one, as it has been inactive for a while and there are few other more specific API-related tickets around.
